### ISSUE TITLE
fit() doesn't end properly (no best pipeline)

### ISSUE THREAD

**Issue by korcky** [rhiever/tpot#499]
fit() method don't even trying find best pipeline.

## Context of the issue

I'm trying to find parameters for my ensemble regressor, but when number of models for ensemble increase fit() method doesn't end properly - ending but don't even trying find best pipeline.

## Process to reproduce the issue

I'm using custom config dictionary, like this:
```
config_dict = {
    'ensemble.regressor.Ensemble': {
        'model1': {
            'ensemble.models.model1.Model1': {
                # parameters
            }
        }

        ...

        'model10': {
            'ensemble.models.model10.Model10': {
                # parameters
            }
        }
}
```
Then:
Step 1. `tpot = TPOTRegressor(..., config_dict=config_dict)`
Step 2. `tpot.fit(X_train, y_train)`
Step 3. `print(tpot.score(X_test, y_test))`

## Expected result

After step 2: best pipeline parameters
After step 3: score of best pipeline

## Current result

After step 2: nothing
After step 3:
```
RuntimeError                              Traceback (most recent call last)
<ipython-input-36-2ae990681f1a> in <module>()
----> 1 print(tpot.score(X_test, y_test))

/usr/local/lib/python3.5/dist-packages/tpot/base.py in score(self, testing_features, testing_target)
    660         """
    661         if self.fitted_pipeline_ is None:
--> 662             raise RuntimeError('A pipeline has not yet been optimized. Please call fit() first.')
    663 
    664         # If the scoring function is a string, we must adjust to use the sklearn

RuntimeError: A pipeline has not yet been optimized. Please call fit() first.
```

## Note

For this testing used dummy ensemble (simple regression where models of ensemble not used) and if delete all parameters of models (`# parameters`) in `config_dict`, or using only 2-3 models with parameters, then step 2 and 3 work correctly.
In tpot exist hidden max number of parameters or why this could happen?

**Comment by weixuanfu2016** [comment]
A nice way to use TPOT. `config_dict` allows this kind of nested estimator as shown in our default [`config_dict`](https://github.com/rhiever/tpot/blob/master/tpot/config/regressor.py#L177-L185)  and we do not limit max number of parameters. Could you please use `verbosity=3` in `tpot = TPOTRegressor(..., config_dict=config_dict)` to check more detailed error message?

**Comment by korcky** [comment]
Nice way to use such a nice thing. With `verbosity=3`:
```
tpot = TPOTRegressor(generations=5, population_size=10, verbosity=3,
                     n_jobs=-1, cv=5, config_dict=tpot_config)
tpot.fit(X_train, y_train)
```
Output:
```
1 operators have been imported by TPOT.
_pre_test decorator: _generate: num_test=0 'Model1'
_pre_test decorator: _generate: num_test=1 'Model1'
_pre_test decorator: _generate: num_test=2 'Model7'
_pre_test decorator: _generate: num_test=3 'Model1'
_pre_test decorator: _generate: num_test=4 'Model8'
_pre_test decorator: _generate: num_test=5 'Model8'
_pre_test decorator: _generate: num_test=6 'Model8'
_pre_test decorator: _generate: num_test=7 'Model7'
_pre_test decorator: _generate: num_test=8 'Model7'
_pre_test decorator: _generate: num_test=9 'Model9'
_pre_test decorator: _generate: num_test=0 'Model9'
_pre_test decorator: _generate: num_test=1 'Model8'
_pre_test decorator: _generate: num_test=0 'Model10'
_pre_test decorator: _generate: num_test=1 'Model8'
_pre_test decorator: _generate: num_test=2 'Model7'
_pre_test decorator: _generate: num_test=3 'Model10'
_pre_test decorator: _generate: num_test=4 'Model9'
_pre_test decorator: _generate: num_test=5 'Model7'
_pre_test decorator: _generate: num_test=6 'Model8'
_pre_test decorator: _generate: num_test=0 'Model8'
_pre_test decorator: _generate: num_test=1 'Model1'
_pre_test decorator: _generate: num_test=2 'Model8'
_pre_test decorator: _generate: num_test=3 'Model8'
_pre_test decorator: _generate: num_test=4 'Model8'
_pre_test decorator: _generate: num_test=5 'Model1'
_pre_test decorator: _generate: num_test=6 'Model7'
_pre_test decorator: _generate: num_test=7 'Model9'
_pre_test decorator: _generate: num_test=8 'Model8'
_pre_test decorator: _generate: num_test=9 'Model8'
_pre_test decorator: _generate: num_test=0 'Model8'
_pre_test decorator: _generate: num_test=1 'Model8'
_pre_test decorator: _generate: num_test=2 'Model9'
_pre_test decorator: _generate: num_test=3 'Model1'
_pre_test decorator: _generate: num_test=4 'Model1'
_pre_test decorator: _generate: num_test=0 'Model8'
_pre_test decorator: _generate: num_test=1 'Model8'
_pre_test decorator: _generate: num_test=2 'Model8'
_pre_test decorator: _generate: num_test=0 'Model8'
_pre_test decorator: _generate: num_test=0 'Model8'
_pre_test decorator: _generate: num_test=1 'Model1'
_pre_test decorator: _generate: num_test=2 'Model7'
_pre_test decorator: _generate: num_test=3 'Model10'
_pre_test decorator: _generate: num_test=0 'Model1'
_pre_test decorator: _generate: num_test=1 'Model8'
_pre_test decorator: _generate: num_test=2 'Model8'
_pre_test decorator: _generate: num_test=3 'Model1'
_pre_test decorator: _generate: num_test=4 'Model1'
_pre_test decorator: _generate: num_test=5 'Model1'
_pre_test decorator: _generate: num_test=6 'Model9'
_pre_test decorator: _generate: num_test=7 'Model9'
_pre_test decorator: _generate: num_test=8 'Model1'
_pre_test decorator: _generate: num_test=9 'Model9'
```

**Comment by weixuanfu2016** [comment]
Hmm, it seems that `_pre_test decorator` printed out too many error messages. But I don't understand why these error messages were "Model*". Also, it seems that some pipelines passed `_pre_test` and should be passed to evaluation step to get CV scores. But no stdout showed up?

Please try the codes below to 
```
# 1 generation is OK
tpot_obj = TPOTRegressor(generations=1, population_size=10, verbosity=3,
                     n_jobs=-1, cv=5, config_dict=tpot_config)
tpot_obj.fit(X_train, y_train)


import numpy as np
from deap import creator
from sklearn.model_selection import cross_val_score
from tpot.export_utils import generate_pipeline_code

# print part of pipeline dictionary
print(dict(list(tpot.evaluated_individuals_.items())[0:2]))
# print a pipeline and its values
pipeline_str = list(tpot.evaluated_individuals_.keys())[0]
print(pipeline_str)
print(tpot.evaluated_individuals_[pipeline_str])
for pipeline_string in sorted(tpot_obj.evaluated_individuals_.keys()):
    # convert pipeline string to scikit-learn pipeline object
    deap_pipeline = creator.Individual.from_string(pipeline_string, tpot_obj._pset)
    sklearn_pipeline = tpot_obj._toolbox.compile(expr=deap_pipeline)
    # print sklearn pipeline string
    sklearn_pipeline_str = generate_pipeline_code(expr_to_tree(individual, tpot_obj._pset), tpot_obj.operators)
    print(sklearn_pipeline_str)
    # Fix random state when the operator allows  (optional) just for get consistent CV score 
    tpot_obj._set_param_recursive(sklearn_pipeline.steps, 'random_state', 42)

    try:
        cv_scores = cross_val_score(sklearn_pipeline, training_features, training_target, cv=5, scoring='accuracy', verbose=0)
        mean_cv_scores = np.mean(cv_scores)
    except Exception as e:
        print(e)
        mean_cv_scores = -float('inf')
    print(mean_cv_scores)

**Comment by korcky** [comment]
Doesn't work: `tpot.evaluated_individuals_` is empty.

**Comment by weixuanfu2016** [comment]
Hmm, it means that no valid pipeline can be randomly generated in the founder population (generation 0) or no pipeline has been a valid CV score

I changed that codes a little bit as below to check if these pipelines are generated and valid:

```
from sklearn.datasets import make_classification
from tpot import TPOTRegressor
pretest_X_reg, pretest_y_reg = make_regression(n_samples=50, n_features=10, random_state=42)
tpot_obj = TPOTRegressor(generations=1, population_size=10, verbosity=3,
                     n_jobs=-1, cv=5, config_dict=tpot_config)
pop = tpot_obj._toolbox.population(n=self.population_size)
for deap_pipeline in pop:
    # convert pipeline string to scikit-learn pipeline object
    sklearn_pipeline = tpot_obj._toolbox.compile(expr=deap_pipeline)
    # print sklearn pipeline string
    sklearn_pipeline_str = generate_pipeline_code(expr_to_tree(deap_pipeline, tpot_obj._pset), tpot_obj.operators)
    print(sklearn_pipeline_str)
    # Fix random state when the operator allows  (optional) just for get consistent CV score 
    tpot_obj._set_param_recursive(sklearn_pipeline.steps, 'random_state', 42)

    try:
        cv_scores = cross_val_score(sklearn_pipeline, training_features, training_target, cv=5, scoring='accuracy', verbose=0)
        mean_cv_scores = np.mean(cv_scores)
    except Exception as e:
        print(e)
        mean_cv_scores = -float('inf')
    print(mean_cv_scores)
"""

**Comment by weixuanfu2016** [comment]
Hmm, it means that no valid pipeline can be randomly generated in the founder population (generation 0) or no pipeline has been a valid CV score

I changed that codes a little bit as below to check if these pipelines are generated and valid:

```
from sklearn.datasets import make_classification
from tpot import TPOTRegressor
pretest_X_reg, pretest_y_reg = make_regression(n_samples=50, n_features=10, random_state=42)
tpot_obj = TPOTRegressor(generations=1, population_size=10, verbosity=3,
                     n_jobs=-1, cv=5, config_dict=tpot_config)
pop = tpot_obj._toolbox.population(n=10)
for deap_pipeline in pop:
    # convert pipeline string to scikit-learn pipeline object
    sklearn_pipeline = tpot_obj._toolbox.compile(expr=deap_pipeline)
    # print sklearn pipeline string
    sklearn_pipeline_str = generate_pipeline_code(expr_to_tree(deap_pipeline, tpot_obj._pset), tpot_obj.operators)
    print(sklearn_pipeline_str)
    # Fix random state when the operator allows  (optional) just for get consistent CV score 
    tpot_obj._set_param_recursive(sklearn_pipeline.steps, 'random_state', 42)

    try:
        cv_scores = cross_val_score(sklearn_pipeline, training_features, training_target, cv=5, scoring='accuracy', verbose=0)
        mean_cv_scores = np.mean(cv_scores)
    except Exception as e:
        print(e)
        mean_cv_scores = -float('inf')
    print(mean_cv_scores)
```

**Comment by korcky** [comment]
Got this (`...` - just a lot of parameters):
```
_pre_test decorator: _generate: num_test=0 'Model10'
_pre_test decorator: _generate: num_test=0 'Model9'
_pre_test decorator: _generate: num_test=1 'Model10'
_pre_test decorator: _generate: num_test=2 'Model10'
_pre_test decorator: _generate: num_test=3 'Model1'
_pre_test decorator: _generate: num_test=4 'Model9'
_pre_test decorator: _generate: num_test=5 'Model8'
_pre_test decorator: _generate: num_test=6 'Model8'
_pre_test decorator: _generate: num_test=0 'Model9'
_pre_test decorator: _generate: num_test=1 'Model10'
_pre_test decorator: _generate: num_test=2 'Model8'
_pre_test decorator: _generate: num_test=3 'Model8'
_pre_test decorator: _generate: num_test=0 'Model1'
_pre_test decorator: _generate: num_test=1 'Model7'
_pre_test decorator: _generate: num_test=2 'Model1'
_pre_test decorator: _generate: num_test=3 'Model1'
_pre_test decorator: _generate: num_test=4 'Model1'
_pre_test decorator: _generate: num_test=5 'Model7'
_pre_test decorator: _generate: num_test=6 'Model9'
_pre_test decorator: _generate: num_test=0 'Model1'
_pre_test decorator: _generate: num_test=1 'Model8'
_pre_test decorator: _generate: num_test=2 'Model3'
_pre_test decorator: _generate: num_test=3 'Model8'
_pre_test decorator: _generate: num_test=4 'Model9'
_pre_test decorator: _generate: num_test=5 'Model1'
_pre_test decorator: _generate: num_test=6 'Model8'
_pre_test decorator: _generate: num_test=7 'Model8'
_pre_test decorator: _generate: num_test=8 'Model8'
_pre_test decorator: _generate: num_test=9 'Model9'
_pre_test decorator: _generate: num_test=0 'Model7'
_pre_test decorator: _generate: num_test=0 'Model9'
_pre_test decorator: _generate: num_test=0 'Model8'
_pre_test decorator: _generate: num_test=1 'Model8'
_pre_test decorator: _generate: num_test=2 'Model8'
_pre_test decorator: _generate: num_test=3 'Model10'
_pre_test decorator: _generate: num_test=0 'Model8'
_pre_test decorator: _generate: num_test=1 'Model8'
_pre_test decorator: _generate: num_test=2 'Model8'
_pre_test decorator: _generate: num_test=3 'Model7'
_pre_test decorator: _generate: num_test=4 'Model1'
_pre_test decorator: _generate: num_test=5 'Model1'
_pre_test decorator: _generate: num_test=6 'Model1'
_pre_test decorator: _generate: num_test=7 'Model8'
_pre_test decorator: _generate: num_test=8 'Model1'
_pre_test decorator: _generate: num_test=9 'Model1'
make_pipeline(
    Ensemble(...)
)
name 'training_features' is not defined
-inf
make_pipeline(
    make_union(
        FunctionTransformer(copy),
        FunctionTransformer(copy)
    ),
    Ensemble(...)
)
name 'training_features' is not defined
-inf
make_pipeline(
    Ensemble(...)
)
name 'training_features' is not defined
-inf
make_pipeline(
    make_union(
        FunctionTransformer(copy),
        FunctionTransformer(copy)
    ),
    Ensemble(...)
)
name 'training_features' is not defined
-inf
make_pipeline(
    Ensemble(...)
)
name 'training_features' is not defined
-inf
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-42-de3937e51b24> in <module>()
      2 for deap_pipeline in pop:
      3     # convert pipeline string to scikit-learn pipeline object
----> 4     sklearn_pipeline = tpot_obj._toolbox.compile(expr=deap_pipeline)
      5     # print sklearn pipeline string
      6     sklearn_pipeline_str = generate_pipeline_code(expr_to_tree(deap_pipeline, tpot_obj._pset), tpot_obj.operators)

/usr/local/lib/python3.5/dist-packages/tpot/base.py in _compile_to_sklearn(self, expr)
    779         sklearn_pipeline: sklearn.pipeline.Pipeline
    780         """
--> 781         sklearn_pipeline = generate_pipeline_code(expr_to_tree(expr, self._pset), self.operators)
    782         return eval(sklearn_pipeline, self.operators_context)
    783 

/usr/local/lib/python3.5/dist-packages/tpot/export_utils.py in generate_pipeline_code(pipeline_tree, operators)
    254 
    255     """
--> 256     steps = _process_operator(pipeline_tree, operators)
    257     pipeline_text = "make_pipeline(\n{STEPS}\n)".format(STEPS=_indent(",\n".join(steps), 4))
    258     return pipeline_text

/usr/local/lib/python3.5/dist-packages/tpot/export_utils.py in _process_operator(operator, operators, depth)
    309             )
    310         else:
--> 311             steps.append(tpot_op.export(*args))
    312     return steps
    313 

/usr/local/lib/python3.5/dist-packages/tpot/operator_utils.py in export(cls, *args)
    252                         arg_value = dep_op_str
    253                     else:
--> 254                         arg_value = "{}({})".format(dep_op_str, ", ".join(dep_op_arguments[dep_op_str]))
    255                     tmp_op_args.append("{}={}".format(dep_op_pname, arg_value))
    256             op_arguments = tmp_op_args + op_arguments

KeyError: 'Model9'
```

**Comment by weixuanfu2016** [comment]
Sorry, I changed `training_features` and `training_target` to `pretest_X_reg` and `pretest_y_reg` in the codes. Could you please try it again?

**Comment by weixuanfu2016** [comment]
And not fit() needed.

**Comment by weixuanfu2016** [comment]
Oh, just notice the scoring function for regression should be `neg_mean_squared_error `. I just updated codes in the comments above
cv_scores = cross_val_score(sklearn_pipeline, pretest_X_reg, pretest_y_reg, cv=5, scoring='neg_mean_squared_error', verbose=0)

**Comment by korcky** [comment]
Negative score and, again, can't find one of the models:
```
1 operators have been imported by TPOT.
_pre_test decorator: _generate: num_test=0 'Model9'
_pre_test decorator: _generate: num_test=1 'Model8'
_pre_test decorator: _generate: num_test=2 'Model1'
...
_pre_test decorator: _generate: num_test=0 'Model9'
_pre_test decorator: _generate: num_test=1 'Model8'
make_pipeline(
    Ensemble(...)
)
-23845.7731237
make_pipeline(
    StackingEstimator(estimator=Ensemble(...)),
    TpotBoosting(...)
)
-23910.755966
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-61-5e259a95c5bc> in <module>()
      5 for deap_pipeline in pop:
      6     # convert pipeline string to scikit-learn pipeline object
----> 7     sklearn_pipeline = tpot_obj._toolbox.compile(expr=deap_pipeline)
      8     # print sklearn pipeline string
      9     sklearn_pipeline_str = generate_pipeline_code(expr_to_tree(deap_pipeline, tpot_obj._pset), tpot_obj.operators)

/usr/local/lib/python3.5/dist-packages/tpot/base.py in _compile_to_sklearn(self, expr)
    779         sklearn_pipeline: sklearn.pipeline.Pipeline
    780         """
--> 781         sklearn_pipeline = generate_pipeline_code(expr_to_tree(expr, self._pset), self.operators)
    782         return eval(sklearn_pipeline, self.operators_context)
    783 

/usr/local/lib/python3.5/dist-packages/tpot/export_utils.py in generate_pipeline_code(pipeline_tree, operators)
    254 
    255     """
--> 256     steps = _process_operator(pipeline_tree, operators)
    257     pipeline_text = "make_pipeline(\n{STEPS}\n)".format(STEPS=_indent(",\n".join(steps), 4))
    258     return pipeline_text

/usr/local/lib/python3.5/dist-packages/tpot/export_utils.py in _process_operator(operator, operators, depth)
    297 
    298         if input_name != 'input_matrix':
--> 299             steps.extend(_process_operator(input_name, operators, depth + 1))
    300 
    301         # If the step is an estimator and is not the last step then we must

/usr/local/lib/python3.5/dist-packages/tpot/export_utils.py in _process_operator(operator, operators, depth)
    306             steps.append(
    307                 "StackingEstimator(estimator={})".
--> 308                 format(tpot_op.export(*args))
    309             )
    310         else:

/usr/local/lib/python3.5/dist-packages/tpot/operator_utils.py in export(cls, *args)
    252                         arg_value = dep_op_str
    253                     else:
--> 254                         arg_value = "{}({})".format(dep_op_str, ", ".join(dep_op_arguments[dep_op_str]))
    255                     tmp_op_args.append("{}={}".format(dep_op_pname, arg_value))
    256             op_arguments = tmp_op_args + op_arguments

KeyError: 'Model8'
```

**Comment by weixuanfu2016** [comment]
I figure out that the `KeyError` is due to a bug about `DEFAULT` value in these parameters of these stacked estimators in Ensemble. I just made a PR to fix this bug. 

Before this PR is merged to dev branch or master branch. You may try my branch for testing if it works in your Ensemble method. The command below may help you to install the branch. Please let me know if this PR fix your issue.

```
pip install --upgrade --no-deps --force-reinstall git+https://github.com/weixuanfu2016/tpot.git@issue499```

**Comment by korcky** [comment]
Score is still negative (in your code) but fit() end properly, score in `print(tpot.score(X_test, y_test))` correct, and with `verbosity=3` no lines like this:
```
_pre_test decorator: _generate: num_test=0 'Model9'
_pre_test decorator: _generate: num_test=2 'Model1'
...
_pre_test decorator: _generate: num_test=0 'Model9'
_pre_test decorator: _generate: num_test=1 'Model8'
```
I think that now all works. Thank you!

**Comment by rhiever** [comment]
PR is merged into the dev branch now.
